libs/payments/paypal/src/lib/paypal.service.ts

```diff
@@ -3,9 +3,57 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 import { Injectable } from '@nestjs/common';
+import { PayPalClient } from './paypal.client';
 import { PayPalManager } from './paypal.manager';
+import { PaypalCustomerManager } from './paypalCustomer/paypalCustomer.manager';
+import { CreatePaypalCustomer } from './paypalCustomer/paypalCustomer.types';
+import { CurrencyManager } from '@fxa/payments/currency';
 
 @Injectable()
 export class PayPalService {
-  constructor(private paypalManager: PayPalManager) {}
+  constructor(
+    private paypalClient: PayPalClient,
+    private paypalManager: PayPalManager,
+    private paypalCustomerManager: PaypalCustomerManager,
+    private currencyManager: CurrencyManager
+  ) {}
+
+  /**
+   * Create and verify a billing agreement is funded from the appropriate
+   * country given the currency of the billing agreement.
+   */
+  async createBillingAgreement(options: {
+    uid: string;
+    token: string;
+    currency: string;
+  }) {
+    const { uid, currency } = options;
+
+    const billingAgreement = await this.paypalClient.createBillingAgreement(
+      options
+    );
+
+    const billingAgreementId = billingAgreement.BILLINGAGREEMENTID;
+
+    const agreementDetails = await this.paypalManager.getBillingAgreement(
+      billingAgreementId
+    );
+
+    const country = agreementDetails.countryCode;
+
+    if (
+      this.currencyManager.isCurrencyCompatibleWithCountry(currency, country)
+    ) {
```

Comment:
I think we'd probably short circuit here and throw an error, since we wouldn't want to proceed if currency is incompatible.

Reply:
What's the best way to do that? I put the throw in `this.currencyManager.isCurrencyCompatibleWithCountry(currency, country) for the mismatch`.

Reply:
Ah, a couple of things here then:
1. A method that's named starting with `is` implies that the method will return `true` or `false` - I wouldn't expect a method with `is` to throw on false.
2. If you want to have isCurrencyCompatibleWithCountry throw, I'd rename it to assertCurrencyCompatibleWithCountry - that's the usual naming to imply that it'll throw an error if false.
3. As it stands, since the method currently functions like an `assert`, you don't need it in the `if` block at all, since the `if` block doesn't really do anything - just have the call to isCurrencyCompatibleWithCountry before the next line with no if block and it'll exit early due to the thrown error.

Reply:
Ah! Got it! Thank you!

Reply:
Done!